src/index.ts

```diff
@@ -137,54 +139,71 @@ class CruxPayPeer {
     }
 
     public async init() {
-        let configService;
-
+        let configService: BlockstackConfigurationService;
         if (this._hasPayIDClaimStored()) {
-            const payIDClaim = this._storage.getJSON("payIDClaim");
+            log.debug("using the stored payIDClaim");
+            const payIDClaim = (this._storage.getJSON("payIDClaim") as ICruxPayClaim);
             // log.debug(`Local payIDClaim:`, payIDClaim)
-            this._setPayIDClaim(new PayIDClaim(payIDClaim as ICruxPayClaim, { getEncryptionKey: this._getEncryptionKey }));
-            const ns: blockstackService.BlockstackService = new blockstackService.BlockstackService({domain: this.walletClientName + CRUX_DOMAIN_SUFFIX});
-            await (this._payIDClaim as PayIDClaim).decrypt();
-            await ns.restoreIdentity((this._payIDClaim as PayIDClaim).virtualAddress as string, {secrets: (this._payIDClaim as PayIDClaim).identitySecrets});
-            const status = await ns.getRegistrationStatus({secrets: (this._payIDClaim as PayIDClaim).identitySecrets});
-            await (this._payIDClaim as PayIDClaim).encrypt();
-            if (status.status === blockstackService.SubdomainRegistrationStatus.DONE) {
-                configService = new BlockstackConfigurationService(this.walletClientName, (this._payIDClaim as PayIDClaim).virtualAddress);
-            } else {
-                configService = new BlockstackConfigurationService(this.walletClientName);
+
+            // if a keyPair is provided, add a validation check on the cruxID stored
+            if (this._keyPair) {
+                const CruxIDArray = await getCruxIDByAddress(config.BLOCKSTACK.BNS_NODES[0], this.walletClientName, this._keyPair.address);
+                if (!CruxIDArray.includes(payIDClaim.virtualAddress as string)) {
+                    throw errors.ErrorHelper.getPackageError(errors.PackageErrorCode.KeyPairMismatch);
+                }
             }
+            this._setPayIDClaim(new PayIDClaim(payIDClaim, { getEncryptionKey: this._getEncryptionKey }));
+            configService = await this._getConfigService();
             await this._initializeNameService(configService);
             await this._restoreIdentity();
+        } else if (this._keyPair) {
+            log.debug("using the keyPair provided");
+            const cruxIDArray = await getCruxIDByAddress(config.BLOCKSTACK.BNS_NODES[0], this.walletClientName, this._keyPair.address);
+            // Assuming the first identity using the walletClientName as the identity
+            if (cruxIDArray[0]) {
+                const payIDClaim = {identitySecrets: {identityKeyPair: this._keyPair}, virtualAddress: (cruxIDArray && cruxIDArray[0]) || undefined};
+                this._setPayIDClaim(new PayIDClaim(payIDClaim, { getEncryptionKey: this._getEncryptionKey }));
+                configService = await this._getConfigService();
+                await this._initializeNameService(configService);
```

Comment:
If this is common across all if, else if statements then we can keep it outside. Not sure in which case it should differ.

Reply:
it differs in the ordering of `await this._initializeNameService(configService);` and ` await this._restoreIdentity();`